Code/GraphMol/RascalMCES/RascalMCES.h

```diff
@@ -0,0 +1,70 @@
+//
+// Copyright (C) David Cosgrove 2023
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+
+#include <RDGeneral/export.h>
+#ifndef RDKIT_RASCAL_MCES_H
+#define RDKIT_RASCAL_MCES_H
+
+#include <vector>
+
+#include <GraphMol/RascalMCES/RascalClusterOptions.h>
+#include <GraphMol/RascalMCES/RascalOptions.h>
+#include <GraphMol/RascalMCES/RascalResult.h>
+namespace RDKit {
+class ROMol;
+
+namespace RascalMCES {
+
+// Find one or more MCESs between the two molecules.  The MCES is the
+// Maximum Common Edge Substructure, and is the largest set of bonds
+// common to the 2 molecules.
+/*!
+ *
+ * @param mol1 : first molecule
+ * @param mol2 : second molecule for MCES determination.
+ * @param opts : (optional) set of options controlling the MCES determination
+ * @return : vector of RascalResult objects.
+ */
+RDKIT_RASCALMCES_EXPORT std::vector<RascalResult> rascalMces(
+    const ROMol &mol1, const ROMol &mol2,
+    const RascalOptions &opts = RascalOptions());
+
+// Cluster the molecules using the Johnson similarity from rascalMces
+// and the algorithm of
+// 'A Line Graph Algorithm for Clustering Chemical Structures Based
+// on Common Substructural Cores', JW Raymond, PW Willett.
+// https://match.pmf.kg.ac.rs/electronic_versions/Match48/match48_197-207.pdf
+// https://eprints.whiterose.ac.uk/77598/
+// This is a fuzzy clustering algorithm, so a molecule may appear in more than
+// one cluster.
+/*!
+ *
+ * @param mols : molecules to cluster
+ * @param clusOpts : (optional) cluster options
+ * @return clusters as vector of vectors of molecules
+ */
+RDKIT_RASCALMCES_EXPORT std::vector<std::vector<std::shared_ptr<ROMol>>>
```

Comment:
It seems like this would be more efficient (and avoid the requirement of constructing shared pointers to use the clustering algorithm) if you just returned `std::vector<std::vector<unsigned int>>` where the ints are the indices of the molecules in each cluster.

Reply:
Do you also think the input should be a vector of plain pointers?  I think that's implied by the comment.  I never use bare pointers these days.  Maybe 2 overloaded functions, one with shared pointers, one with bare ones?

Reply:
I think using bare pointers to functions which are just going to work with whatever they point to (not take/share ownership) is fine. It also can simplify code and speed things up a bit - if you don't already have shared pointers then you don't need to make copies so that you can create them.